packages/playwright-core/src/server/fetch.ts

```diff
@@ -307,8 +309,27 @@ export abstract class APIRequestContext extends SdkObject {
       // If we have a proxy agent already, do not override it.
       const agent = options.agent || (url.protocol === 'https:' ? httpsHappyEyeballsAgent : httpHappyEyeballsAgent);
       const requestOptions = { ...options, agent };
+
+      const startAt = monotonicTime();
+      let dnsLookupAt: number | undefined;
+      let tcpConnectionAt: number | undefined;
+      let tlsHandshakeAt: number | undefined;
+      let requestFinishAt: number | undefined;
+      let firstByteAt: number | undefined;
+      let endAt: number | undefined;
+
       const request = requestConstructor(url, requestOptions as any, async response => {
         const notifyRequestFinished = (body?: Buffer) => {
+          const timings: har.Timings = {
+            send: requestFinishAt! - startAt,
```

Comment:
I see this method called multiple times, how are we sure that `requestFinishAt` has been set in all of them?

Reply:
`notifyRequestFinished` is called exclusively inside the callback of `requestConstructor`, which is a listener for the `response` event. My reading of the Node.js docs is that the `finish` event (which sets `requestFinishAt`) is guaranteed to fire before the `response` event.